subsys/logging/log_backend_adsp_mtrace.c

```diff
@@ -0,0 +1,252 @@
+/*
+ * Copyright (c) 2019,2022 Intel Corporation
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <zephyr/logging/log_backend_adsp_mtrace.h>
+#include <zephyr/logging/log_backend.h>
+#include <zephyr/logging/log_core.h>
+#include <zephyr/logging/log_output.h>
+#include <zephyr/logging/log_backend_std.h>
+#include <zephyr/zephyr.h>
+#include <soc.h>
+#include <adsp_memory.h>
+
+/*
+ * A lock is needed as log_process() and log_panic() have no internal locks
+ * to prevent concurrency. Meaning if log_process is called after
+ * log_panic was called previously, log_process may happen from another
+ * CPU and calling context than the log processing thread running in
+ * the background. On an SMP system this is a race.
+ *
+ * This caused a race on the output trace such that the logging output
+ * was garbled and useless.
+ */
+static struct k_spinlock mtrace_lock;
+
+static uint32_t log_format_current = CONFIG_LOG_BACKEND_ADSP_MTRACE_OUTPUT_DEFAULT;
+
+static adsp_mtrace_log_hook_t mtrace_hook;
+
+static bool mtrace_active;
+
+/*
+ * SRAM window for mtrace is organized in 16 (equal sized) pages:
+ *
+ *	------------------------
+ *	| Page0  - descriptors |
+ *	------------------------
+ *	| Page1  - slot0       |
+ *	------------------------
+ *	| Page2  - slot1       |
+ *	------------------------
+ *	|         ...          |
+ *	------------------------
+ *	| Page14  - slot13     |
+ *	------------------------
+ *	| Page15  - slot14     |
+ *	------------------------
```

Comment:
Another design note: what's the point to having different "slots" for output?  Absent other metadata (like timestamps or sequence numbers), that has the effect of losing the time order of events, which can be critically important when reconstructing a failure.  The logging API already allows for filtering by subsystem, etc...  What is it that you're trying to segregate by using "slots", and do you really need that?  Note also that having lots of much smaller streams means that a big dump from any one of them is much more likely to overflow due to buffer size.

Reply:
Actually coming back, it seems like this is just an 8k window, so the number of pages is just two, and so there's only one slot in practice?  Might be worth calling out in the docs here, because seeing 64k of memory defined sorta scared me :)

Reply:
And even more whining: if the only purpose of that initial 4k slot 0 page is to store 2-3 words of data for one slot... maybe find a way to put it somewhere else?  It could trivially be appended to the FW_STATUS window (itself already very large).

Reply:
See my previous comment. In short, this is the protocol as it is defined. The number of slots can vary depending on how the FW sets up the windows in boot.c. We are currently limited to CONFIG_HP_SRAM_RESERVE, but in theory more of the SRAM could be reserved for the windows (but in practise not).

I fill the descriptors with same data used to set up the windows in boot.c, so the information reflects actual programming of the windows.